File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/letter_combinations_of_a_phone_number/LetterCombinationsofaPhoneNumber.java
package main.java.com.jiang.leetcode100.letter_combinations_of_a_phone_number;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Solution {
    public List<String> letterCombinations(String digits) {
        
        String[] mapping = new String[]{
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
        };

        List<String> result = new ArrayList<>();
        if(!digits.isEmpty()) result.add("");
        for(int i = 0; i < digits.length(); i++) {
            List<String> temp = new ArrayList<>();
            String cur_chars = mapping[digits.charAt(i) - '2'];
            for(String s : result) {
                for(int j = 0; j < cur_chars.length(); j++) {
                    temp.add(s + cur_chars.charAt(j));
                }
            }
            result.clear();
            result.addAll(temp);
        }

        return result;
    }
}

public class LetterCombinationsofaPhoneNumber {
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String digits = scanner.nextLine();
        scanner.close();

        List<String> result = new Solution().letterCombinations(digits);
        for(String s : result) {
            System.out.println(s);
        }
    }

}

File: Leetcode100/leetcode100_python/LetterCombinationsofaPhoneNumber.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        mapping = {
            '2': "abc",
            '3': "def",
            '4': "ghi",
            '5': "jkl",
            '6': "mno",
            '7': "pqrs",
            '8': "tuv",
            '9': "wxyz"}
        result = [""]
        for digit in digits:
            temp = []
            for s in result:
                for ch in mapping[digit]:
                    temp.append(s + ch)
            result = temp
        return result

if __name__ == "__main__":
    digits = input()
    solution = Solution()
    print(solution.letterCombinations(digits))
        